File: Sources/MyATSTypesetter.m
#import "MyATSTypesetter.h"
#import "MyLayoutManager.h"

@implementation MyATSTypesetter
-(NSTypesetterControlCharacterAction)actionForControlCharacterAtIndex:(NSUInteger)charIndex
{
    NSTypesetterControlCharacterAction action = [super actionForControlCharacterAtIndex:charIndex];
    
    if (action & NSTypesetterZeroAdvancementAction) {
        unichar character = [self.attributedString.string characterAtIndex:charIndex];
        if (!CFStringIsSurrogateLowCharacter(character)) {
            return NSTypesetterWhitespaceAction;
        }
    }
    
    return action;
}

-(NSRect)boundingBoxForControlGlyphAtIndex:(NSUInteger)glyphIndex
                          forTextContainer:(NSTextContainer*)textContainer
                      proposedLineFragment:(NSRect)proposedRect
                             glyphPosition:(NSPoint)glyphPosition
                            characterIndex:(NSUInteger)charIndex
{
    NSRect rect = proposedRect;
    rect.size.width = ((MyLayoutManager*)(self.layoutManager)).replacementGlyphWidth;
    return rect;
}
@end

File: Sources/MyATSTypesetter.swift
import Foundation

class MyATSTypesetter: NSATSTypesetter {
    override func actionForControlCharacter(at index: Int) -> NSTypesetterControlCharacterAction {
        let action = super.actionForControlCharacter(at: index)
        
        if action.contains(.zeroAdvancementAction),
           let character = (self.attributedString?.string as NSString?)?.character(at: index),
           !CFStringIsSurrogateLowCharacter(character) {
            return .whitespaceAction
        }
    
        return action
    }
    
    override func boundingBox(forControlGlyphAt glyphIndex: Int,
                              for textContainer: NSTextContainer,
                              proposedLineFragment proposedRect: NSRect,
                              glyphPosition: NSPoint,
                              characterIndex charIndex: Int) -> NSRect {
        var rect = proposedRect
        if let width = (self.layoutManager as? MyLayoutManager)?.replacementGlyphWidth() {
            rect.size.width = width
        }
        return rect
    }
}


